src/compiler/scala/tools/nsc/backend/jvm/BytecodeWriters.scala

```diff
@@ -118,10 +135,20 @@ trait BytecodeWriters {
     def writeClass(label: String, jclassName: String, jclassBytes: Array[Byte], outfile: AbstractFile) {
       assert(outfile != null,
              "Precisely this override requires its invoker to hand out a non-null AbstractFile.")
-      val outstream = new DataOutputStream(outfile.bufferedOutput)
+      if (outfile.file != null) {
+        try {
+          Files.write(outfile.file.toPath, jclassBytes)
+        } catch {
+          case _: java.nio.file.NoSuchFileException =>
```

Comment:
I like the approach of caching the results of file stats, but opted not to do that in this PR because it takes a bit of care to get the lifecycle of that cache sorted out (we need to start fresh on a new compiler run, but the current design of `ByteCodeWriters` didn't have a hook for that part of the lifecycle.)